fairscale/nn/data_parallel/fully_sharded_data_parallel.py

```diff
@@ -1354,18 +1350,24 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
                 # Average grad by world_size for consistency with PyTorch DDP.
                 param.grad.data.div_(self.gradient_predivide_factor)
 
-            callback_fn = functools.partial(self._post_reduction_hook, param)
             if param._is_sharded:
-                assert param._is_sharded
                 assert self._reducer is not None
-                grad_chunks = chunk_and_pad(param.grad.data, self.world_size)
+                # Save the unsharded grad for reduction. We will asynchronously accumulate the reduced gradient into
+                # param._saved_grad_shard. If this FSDP module was called multiple times it's possible that multiple
+                # gradient reductions will happen in an undefined order. But addition commutes, so this order doesn't
+                # matter, neglecting rounding.
+                grad = param.grad.data
+                # Clear grad on the tensor, so any repeated gradient computations do not interfere with this reduction.
+                param.grad = None
```

Comment:
This may increase the memory usage a bit due to 2 copies of the grad for the multiple forward pass + checkpointing case, right? Did you see the impact from that? Worth adding a comment about it here?

Reply:
The memory usage should be the same as we had before in the case of traditional gradient accumulation over the batch dimension. In that case, we would have saved the sharded gradient as well. Also if there is only one forward pass, or the reduction always happens fast enough, the memory usage should be unchanged. Only one unsharded gradient tensor would exist at a time. 

I'll take a closer look tomorrow, but I'm expecting a reduction in memory usage vs the forward counters. Keeping one more sharded tensor should be much better than keeping any unsharded tensors, especially with say 8 gpus.

Reply:
The memory usage is definitely reduced with these changes. I'm able to run larger models locally than I could before, without OOM.

I elaborated the comment to discuss memory usage, including a description of how we can mitigate the 2 copy edge case by using CUDA events to limit the number of outstanding reduction operations.